internal/handlers/msg_service_router.go

```diff
@@ -0,0 +1,189 @@
+				// work correctly anyway.
+				panic(fmt.Errorf("can't register request type %T for service method %s", i, fqMethod))
+			}
+
+			requestTypeName = sdk.MsgTypeURL(msg)
+			return nil
+		}, noopInterceptor)
+
+		// Check that the service Msg fully-qualified method name has already
+		// been registered (via RegisterInterfaces). If the user registers a
+		// service without registering according service Msg type, there might be
+		// some unexpected behavior down the road. Since we can't return an error
+		// (`Server.RegisterService` interface restriction) we panic (at startup).
+		reqType, err := msr.interfaceRegistry.Resolve(requestTypeName)
+		if err != nil || reqType == nil {
+			panic(
+				fmt.Errorf(
+					"type_url %s has not been registered yet. "+
+						"Before calling RegisterService, you must register all interfaces by calling the `RegisterInterfaces` "+
+						"method on module.BasicManager. Each module should call `msgservice.RegisterMsgServiceDesc` inside its "+
+						"`RegisterInterfaces` method with the `_Msg_serviceDesc` generated by proto-gen",
+					requestTypeName,
+				),
+			)
+		}
+
+		// Check that each service is only registered once. If a service is
+		// registered more than once, then we should error. Since we can't
+		// return an error (`Server.RegisterService` interface restriction) we
+		// panic (at startup).
+		_, found := msr.routes[requestTypeName]
+		if found {
+			panic(
+				fmt.Errorf(
+					"msg service %s has already been registered. Please make sure to only register each service once. "+
+						"This usually means that there are conflicting modules registering the same msg service",
+					fqMethod,
+				),
+			)
+		}
+
+		msr.routes[requestTypeName] = func(ctx sdk.Context, req sdk.Msg) (*sdk.Result, error) {
+			msgTypeURL := sdk.MsgTypeURL(req)
+
+			feeGasMeter, ok := ctx.GasMeter().(*antewrapper.FeeGasMeter)
+			if !ok {
+				panic("GasMeter is not of type FeeGasMeter")
+			}
+
+			tx, err := msr.decoder(ctx.TxBytes())
+			if err != nil {
+				panic(fmt.Errorf("error msg handling while getting txBytes: %w", err))
+			}
+
+			feeTx, ok := tx.(sdk.FeeTx)
+			if feeTx == nil || !ok {
+				panic("only Fee Tx are supported on provenance.")
+			}
+
+			fee, err := msr.msgBasedFeeKeeper.GetMsgBasedFee(ctx, msgTypeURL)
+			if err != nil {
+				return nil, err
+			}
+			if fee != nil && fee.AdditionalFee.IsPositive() {
+				ctx.Logger().Debug(fmt.Sprintf("Tx Msg %v has an additional fee of %v ", msgTypeURL, fee.AdditionalFee))
+
+				if !feeGasMeter.IsSimulate() {
+					err = antewrapper.EnsureSufficientFees(runtimeGasForMsg(ctx), feeTx.GetFee(), feeGasMeter.FeeConsumed().Add(fee.AdditionalFee),
+						msr.msgBasedFeeKeeper.GetFloorGasPrice(ctx), msr.msgBasedFeeKeeper.GetDefaultFeeDenom())
+					if err != nil {
+						return nil, err
+					}
+				}
+
+				feeGasMeter.ConsumeFee(fee.AdditionalFee, msgTypeURL)
+			}
+			ctx = ctx.WithEventManager(sdk.NewEventManager())
```

Comment:
My mistake not sure if that was a copy and paste error.  Good catch!

Reply:
hold on this is coming from the cosmod sdk code 
```	msr.routes[requestTypeName] = func(ctx sdk.Context, req sdk.Msg) (*sdk.Result, error) {
			ctx = ctx.WithEventManager(sdk.NewEventManager())
			interceptor := func(goCtx context.Context, _ interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
				goCtx = context.WithValue(goCtx, sdk.SdkContextKey, ctx)
				return handler(goCtx, req)
			}
			// Call the method handler from the service description with the handler object.
			// We don't do any decoding here because the decoding was already done.
			res, err := methodHandler(handler, sdk.WrapSDKContext(ctx), noopDecoder, interceptor)
			if err != nil {
				return nil, err
			}

			resMsg, ok := res.(proto.Message)
			if !ok {
				return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "Expecting proto.Message, got %T", resMsg)
			}

			return sdk.WrapServiceResult(ctx, resMsg, err)
		}```

Reply:
Yes, after running the tests without it.  There were lots of duplicate event messages.  For example, after removing it there was 30 events when 15 were expected (authz test was 68 when 20 was expected).  My first response is half correct.  It was a copy and paste of the original sdk code, but it isn't an error.

Reply:
Yes, after running the tests without it. There were lots of duplicate event messages. For example, after removing it there was 30 events when 15 were expected (authz test was 68 when 20 was expected). My first response is half correct. It was a copy and paste of the original sdk code, but it isn't an error.

Reply:
I actually remember that bug and this line was a fix for that now. :doh: